### ISSUE TITLE
ImportError: Compatibility w/recent dask?

### ISSUE THREAD

**Issue by keflavich** [intake/intake-astro#9]
I got an exception on this line:

https://github.com/intake/intake-astro/blob/master/intake_astro/array.py#L52

```python
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
/scratch/local/45915091/ipykernel_32764/2861514433.py in <module>
      1 img_ = FITSArraySource("file.fits", ext=1)
----> 2 img = img_.to_dask()
      3 img

/blue/adamginsburg/adamginsburg/repos/intake-astro/intake_astro/array.py in to_dask(self)
    101 
    102     def to_dask(self):
--> 103         self._get_schema()
    104         return self.arr
    105 

/blue/adamginsburg/adamginsburg/repos/intake-astro/intake_astro/array.py in _get_schema(self)
     50 
     51     def _get_schema(self):
---> 52         from dask.bytes import open_files
     53         import dask.array as da
     54         from dask.base import tokenize

ImportError: cannot import name 'open_files' from 'dask.bytes' (/blue/adamginsburg/adamginsburg/repos/dask/dask/bytes/__init__.py)
```

Has the `open_files` function been renamed?

**Comment by martindurant** [comment]
This is a little ancient! open_files now lives in fsspec, which was forked out of dask some time ago. Note that astropy is currently adding fsspec support for fsspec, to access remote data directly; but there is no explicit dask support anywhere else yet. This is something I also have on the roadmap in the kerchunk.fits module, which is a different thing again. Sprry about this fragmentation!

**Comment by keflavich** [comment]
OK, thanks, good to know I shouldn't be using this.

What's the recommended way to access FITS data at the moment?  Just `dask.array.from_array(fits.getdata(fn))`?

**Comment by martindurant** [comment]
There's probably no point doing from_array(getdata), since then all the data will have been read into  memory already. You could use the trick in this repo to run section() on the fits extensions to grab pieces without a full load - but this needs to be done in workers, not the main process. You could use dleayed() or an explicit dask graph to make a dask array out of this.

I know that ndcube (from sunpy) has some dask integration. Might be easier hooking on to what is there. 

Direct and explicit dask support in astropy is of interest to the community, but I haven't seen much progress.
https://github.com/astropy/astropy-project/pull/283
https://github.com/astropy/astropy-project/pull/275

**Comment by Cadair** [comment]
> I know that ndcube (from sunpy) has some dask integration. Might be easier hooking on to what is there.

It does but you have to get a dask array first, it doesn't have any loader magic for FITS.

**Comment by martindurant** [comment]
Do you @Cadair think it's worth bringing the code here up to date for that purpose then, or can it be upstreamed to ndcube? 

I'm sure you also noticed https://github.com/astropy/astropy/pull/13238 , which is relevant.

**Comment by Cadair** [comment]
Could you remind me what this does vs kerchunk etc in terms of FITS loading? I think ultimately something to do delayed chunked loading belongs in astropy core. (I.e. open a fits file get a dask array back)

**Comment by martindurant** [comment]
Yes, I suppose you can phrase it all in kerchunk, but you need to build the references first, rather than just asserting that a set of FITS files belong together. We don't have a way to do that in one go with kerchunk (although I can see how it would be possible). Of course, the main point here, is to provide an *intake* loader, i.e., a way to declaratively describe datasets. Via kerchunk, you would be doing that with intake-xarray (i.e., the zarr loader).

I totally agree that astropy should allow for lazy, chunkwise IO. At the moment, I can only see that happening with a zarr backend to its IO.
